### ISSUE TITLE
Rule suggestion: no-global-setup

### ISSUE THREAD

**Issue by ganimomer** [tlvince/eslint-plugin-jasmine#29]
This rule would enforce that no `beforeEach`, `afterEach`, `beforeAll`, or `AfterAll` are defined outside a `describe`.

For example:
Invalid:

``` js
beforeEach(function() {
  //Code goes here...
});
describe(function() {
});
```

valid:

``` js
describe(function() {
  beforeEach(function) {
    //Code goes here...
  });
});
```

**Comment by burabure** [comment]
@ganimomer want to give it a go?, if you never worked with eslint rules before, it's a great learning experience. You can start here: http://eslint.org/docs/developer-guide/working-with-rules and also you'll need this https://astexplorer.net/

**Comment by Moeriki** [comment]
Why is no-global-setup recommended as error? I mean, what's the down-side of having global setup. Would it be possible to add this to the docs.

**Comment by ganimomer** [comment]
Using `beforeEach` and `afterEach` etc outside describe would make the callback code run before **all suites in the current test run**, not just the current file.
In most cases, this is not intentional. (And in the rare cases of doing this intentionally, the rule should be ignored/turned off for the file).

I can add it in a PR if you'd like.

**Comment by Moeriki** [comment]
Thanks for explaining! I can see how that could yield unwanted side-effects. Off to fix my tests.

Rules enforcing good practices should definitely have an explanation in the docs.

**Comment by JELaVallee** [comment]
Adding my two cents on this one... I just ran into this on a new project where we setup a global beforeAll in our helpers for injecting common extended matchers that we use throughout our node app's specs.  E.g. using promise-matcher in code that's festooned with ES6-style promises.

Suggestion: Have the rule only applied in files that match the jasmine.json's "specs" collection so it's only applied inside of specs (where that level of before/afterEach/All differentiation is valuable).

**Comment by DianaSuvorova** [comment]
@JELaVallee , could you provide the snippet of the global beforeAll you are using in your project. How is structurally different from an invalid example for this rule ie:

```js
beforeEach(function() {
  //Code goes here...
});
describe(function() {
});
```

**Comment by JELaVallee** [comment]
Sure can!

The context in my current case is a node-express application, so, we're bootstrapping jasmine via helpers.  Helpers to provide customer stdio output, helpers for different report generation for various monitoring systems we have, helpers to provide custom logging for alerts monitoring, etc.

For the sake of my previous comment, let's use the example above of injecting the 'jasmine-node-promise-matchers' module into a global scope...

The "spec-scoped" approach to including said matchers in a test file is to do the following (as documented in their repo: https://github.com/agirorn/jasmine-node-promise-matchers#install):

```javascript
beforeEach(function() {
  jasmine.addMatchers(require('jasmine-node-promise-matchers'));
});
```

This requires that line in all of our spec files' `describe(...beforeEach()...)` blocks -or- at least in a `beforeAll()` block in nested `describe(...describe()*...)` spec-sets.

Our project already has ~100 component modules being covered, each with a *.spec.js file, so you can see how this can become a good bit of code-crud and also an easily overlooked inclusion that can lead to unnecessary hair-pulling.  We also have several matcher modules we're including, as well as a couple custom ones.  More code-crud...

One solution is to bundle all of them up into a single "sack-of-modules" import... Still, ugh.

Our solution has been to create a helper such as this:

In _/spec/helpers/common-matchers-helper.js_ : 

```javascript
const JasminePromiseMatchers = require('jasmine-node-promise-matchers');
// import more matchers
...
beforeEach( () => {        // This get's flagged if no-global-setup is enabled 
    jasmine.addMatchers(JasminePromiseMatchers);
    // apply more matchers to the jasmine context
...
});
```

Our jasmine.json config has `/spec/helpers/**/*.js` in the "helpers" collection, so this will get picked up when jasmine is initialized and then the above is executed/eval'd immediately before the spec file is evaluated.   

If we attempt to just do the `jasmine.addMatchers(...)` w/o the surrounding `beforeEach(...)` then jasmine throws an exception that addMatchers needs to be inside a beforeEach/beforeAll/afterEach/afterAll block to have proper scope.

An ideal solution would be to have a `jasmine.getEnv().addMatchers()` method for applying global matchers outside a before/after scope similar to how custom reporters are applied.  Alas, that's a PR for a different project...

The only other solution I've been able to come up with is to take this same matchers-helper approach but instead of using the `jasmine.addMatchers()` method, just binding the the matcher methods from our various modules directly onto the `jasmine.prototype`.  Now that I'm done puking a little, we can agree to never discuss that as a "solution" ever again.  ;-)

Totally open to suggestions, ideas.  Fire away...

**Comment by JELaVallee** [comment]
Oh yeah, just so you know where I got that from... It's part of the jasmine's examples: https://github.com/jasmine/jasmine/blob/8624a52ee0b6f13b3b608ea6417ccc02257c5412/lib/jasmine-core/example/spec/SpecHelper.js

So, it _is_ not a completely un-natural exception to the rule that 'no-global-setup' is intended to enforce.

**Comment by DianaSuvorova** [comment]
@JELaVallee , thanks a bunch for details explanation. Having global beforeEach/beforeAll makes sense to me. 
Is it be safe to assume that the file that has global jasmine setup wouldn't have any `describe` blocks as part of the file . And the file the has describe blocks shouldn't have a global setup?

**Comment by JELaVallee** [comment]
@DianaSuvorova No problem!  It's a sticky little detail that's not clearly documented in jasmine short of reading the source/examples in detail.

I've forked this repo and gonna see if I can get the unit test/code for no-global-setup to reflect the conditions properly, but it goes like this:

1. Inside a spec file, beforeEach/All and afterEach/All should exist only inside describe blocks since jasmine effectively runs the various spec files all in the same execution context, so an improperly wrapped beforeAll in test1.spec.js could potentially corrupt the state of test2.spec.js downstream.  (This is covered properly by the current rule)
2. Outside a spec file, but inside a "helper" file as defined by the "helpers" file regexp matcher in the jasmine.json config, you can have a beforeEach/All, afterEach/All defined outside a describe block because it's a block you want to always have applied (e.g. for setting global env-vars or injecting reporters/matchers like we do).  (This is not covered by the current rule implementation)

I'll take a crack at applying the logic for #2 above and submit a PR if I get it working well.  It's basically a matter of looking at the bound jasmine config's helpers matcher and checking if the current scanned node's filename matches the pattern, if it does, then don't apply the rule.

**Comment by DianaSuvorova** [comment]
@JELaVallee , oh I see. for the second one -  instead of checking for the filename pattern, I was wondering if we can assume that  if the whole file has 0 `describe` blocks we should skip the rule.

But I could see your approach as well. Looking forward to your PR and LMK if you need help 🙂

**Comment by JELaVallee** [comment]
@DianaSuvorova 

Have a look at #139 

Ended up taking your simplified unit-isolated approach vs. the hell of injecting a mocked jasmine context with helper definitions.

Test and docs updated to reflect as well.

Cheers!
